lib/OpenQA/Settings.pm

```diff
@@ -0,0 +1,55 @@
+# Copyright (C) 2019 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+
+package OpenQA::Settings;
+
+use strict;
+use warnings;
+
+# replace %NAME% with $settings{NAME}
+sub expand_placeholders {
```

Comment:
May I suggest an alternative to the expand_placeholders subroutine?

```
sub expand_placeholders {
    my ($settings) = @_;

    for my $value (sort values %$settings) {
        next unless defined $value;
        my %seen;
        $value =~ s/%(\w+)%/expand_placeholder($settings, $1, \%seen)/ge;

    }
    return 1;
}

sub expand_placeholder {
    my ($settings, $key, $global_seen) = @_;
    unless (exists $settings->{ $key }) {
        return '';
    }
    my %seen = %$global_seen;
    # prevent circular replacement
    if ($seen{ $key }++) {
        die sprintf "The key %s contains a circular reference, its value is %s",
            $key, $settings->{ $key };
        die "Circular ref $key";
    }
    my $value = $settings->{ $key };
    $value =~ s/%(\w+)%/expand_placeholder($settings, $1, \%seen)/ge;
    return $value;
}
```

Reply:
I prefer throwing exceptions too. But i don't think we have an exception class in openQA yet to filter which exceptions to show the user.

Reply:
@perlpunk If I may give a suggestion, start your github code section comments with the word "perl" immediately after the three backticks as I just did with your comment for cool syntax highlighting :)

Reply:
@okurz done, thanks :)

Reply:
@perlpunk Thanks for your suggestion. It looks better.  I would like to ask your opinion @perlpunk @kraih @okurz  Do I need to change `return` to `die`.  If I do not need to do that, I prefer to keep current method `expand_placeholders`.